tidy3d/components/medium.py

```diff
@@ -306,76 +336,283 @@ def from_nk(cls, n: float, k: float, freq: float, **kwargs):
-        """Complex-valued permittivity as a function of frequency."""
+    def eps_dataset_freq(self, frequency: float) -> PermittivityDataset:
+        """Permittivity dataset at ``frequency``. The dispersion comes
+        from DC conductivity that results in nonzero Im[permittivity].
 
-        eps_xx = self.xx.eps_model(frequency)
-        eps_yy = self.yy.eps_model(frequency)
-        eps_zz = self.zz.eps_model(frequency)
-        return np.mean((eps_xx, eps_yy, eps_zz))
+        Parameters
+        ----------
+        frequency : float
+            Frequency to evaluate permittivity at (Hz).
+
+        Returns
+        -------
+        PermittivityDataset
+            The permittivity evaluated at ``frequency``.
+        """
+
+        new_field_components = {}
+        for name, eps_dataset_component in self.eps_dataset.field_components.items():
+            freq = eps_dataset_component.f[0]
+            eps_freq = (
+                eps_dataset_component.real + 1j * eps_dataset_component.imag * freq / frequency
+            )
+            eps_freq = eps_freq.assign_coords({"f": [frequency]})
+            new_field_components.update({name: eps_freq})
+        return PermittivityDataset(**new_field_components)
+
+    def eps_diagonal_on_grid(
+        self,
+        frequency: float,
+        coords: Coords,
+    ) -> Tuple[Numpy, Numpy, Numpy]:
+        """Spatial profile of main diagonal of the complex-valued permittivity
+        at ``frequency`` interpolated at the supplied coordinates.
+
+        Parameters
+        ----------
+        frequency : float
+            Frequency to evaluate permittivity at (Hz).
+        coords : :class:`.Coords`
+            The grid point coordinates over which interpolation is performed.
+
+        Returns
+        -------
+        Tuple[Numpy, Numpy, Numpy]
+            The complex-valued permittivity tensor at ``frequency`` interpolated
+            at the supplied coordinate.
+        """
+
+        eps_freq = self.eps_dataset_freq(frequency)
+        interp_shape = [len(coord_comp) for coord_comp in coords.to_list]
+        eps_list = [
+            np.array(
+                self._interp(eps_freq.field_components[comp], coords, self.interp_method)
+            ).reshape(interp_shape)
+            for comp in ["eps_xx", "eps_yy", "eps_zz"]
+        ]
+        return tuple(eps_list)
 
     @ensure_freq_in_range
     def eps_diagonal(self, frequency: float) -> Tuple[complex, complex, complex]:
-        """Main diagonal of the complex-valued permittivity tensor as a function of frequency."""
+        """Main diagonal of the complex-valued permittivity tensor
+        at ``frequency``. Spatially, we take max{|eps|}, so that autoMesh generation
+        works appropriately.
+        """
+        eps_freq = self.eps_dataset_freq(frequency)
+        eps_np_list = [
+            np.array(eps_freq.field_components[comp]).ravel()
+            for comp in ["eps_xx", "eps_yy", "eps_zz"]
```

Comment:
what happens if any of these fields are missing? Also, might be a bit more simple to do
```py
eps_np_list = [np.array(sclr_fld).ravel() for _, sclr_fld in eps_freq.field_components.items()]
```

Reply:
same general comments at line ~460 above

Reply:
One question here: when unpacked with `.items`, will it be in order (eps_xx, eps_yy, eps__zz)?

Reply:
Should be, yes. Because after python 3.6 or something dictionaries are ordered.

Reply:
> what happens if any of these fields are missing? 

All the fields are required as defined in `PermittivityDataset`: https://github.com/flexcompute/tidy3d/blob/28d74cfc9965fe3aea79a3198fe558c6e2561e47/tidy3d/components/data/dataset.py#L369-L383

Reply:
ah ok.